### PR TITLE
3482 Build component instance from dictionary config

### PR DESCRIPTION
**Pull request by Nic-Ma**
Task 2 of #3482 .

### Description
This PR implemented the utility to build a component instance from dictionary config, searching from specified packages and modules.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_config_parser.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        result = configer.build_component(test_input)
+        if result is not None:
+            self.assertTrue(isinstance(result, output_type))
+            self.assertEqual(result.keys[0], "image")
+        else:
+            # test `disabled` works fine
+            self.assertEqual(result, output_type)
+
+
+@skip_if_windows
```
Comment (rijobro) [inline tests/test_config_parser.py:59]: How come this doesn't work on Windows?

**Review by wyli — CHANGES_REQUESTED** [review]
it seems this is not compatible with the lazy instantiation idea, perhaps we should agree on a proof-of-concept before writing small pull requests.

File: tests/test_config_parser.py (hunk 1 above)
Comment (Nic-Ma) [reply tests/test_config_parser.py:59]: Hi @rijobro ,

This is an error when parsing the paths of PyTorch:
```
ERROR: test_type_3 (tests.test_config_parser.TestConfigParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\parameterized\parameterized.py", line 533, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "D:\a\MONAI\MONAI\tests\test_config_parser.py", line 47, in test_type
    configer = ConfigParser(**input_param)
  File "D:\a\MONAI\MONAI\monai\apps\mmars\config_parser.py", line 68, in __init__
    self.module_scanner = ModuleScanner(pkgs=pkgs, modules=modules)
  File "D:\a\MONAI\MONAI\monai\apps\mmars\config_parser.py", line 34, in __init__
    self._class_table = self._create_classes_table()
  File "D:\a\MONAI\MONAI\monai\apps\mmars\config_parser.py", line 41, in _create_classes_table
    for _, modname, _ in pkgutil.walk_packages(path=package.__path__, prefix=package.__name__ + "."):
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\pkgutil.py", line 107, in walk_packages
    yield from walk_packages(path, info.name+'.', onerror)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\pkgutil.py", line 92, in walk_packages
    __import__(info.name)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\torch\distributed\_sharded_tensor\__init__.py", line 15, in <module>
    from .api import (
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\torch\distributed\_sharded_tensor\api.py", line 145, in <module>
    def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]], rpc_rank: int):
AttributeError: module 'torch.distributed.rpc' has no attribute 'RRef'
```
I will check it deeper.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review and feedback.
Sorry maybe I didn't describe the PR clearly enough, it doesn't conflict with the `lazy instantiation` idea, it's only the utility to build one component from the dictionary.
For the `lazy instantiation` idea, I think it's the logic of handling the **whole config content**, when it's really necessary to create an instance, it will call this utility to create it. I will implement something like `ConfigResolver` to handle the whole config for the step 3 of #3482 later when we discussed with more team guys.

Thanks.

File: tests/test_config_parser.py (hunk 1 above)
Comment (Nic-Ma) [reply tests/test_config_parser.py]: Hi @rijobro ,

The issue had been fixed in the latest commit.

Thanks.

**Comment by wyli** [comment]
> Hi @wyli ,
> 
> Thanks for your review and feedback. Sorry maybe I didn't describe the PR clearly enough, it doesn't conflict with the `lazy instantiation` idea, it's only the utility to build one component from the dictionary. For the `lazy instantiation` idea, I think it's the logic of handling the **whole config content**, when it's really necessary to create an instance, it will call this utility to create it. I will implement something like `ConfigResolver` to handle the whole config for the step 3 of #3482 later when we discussed with more team guys.
> 
> Thanks.

Maybe there's a misunderstanding here, the lazy instantiation will work at the component level as well. Also, it'll be useful to have a complete prototype and then we divide it into smaller pieces and merge. That'll be an efficient way to roll out a major feature.

**Comment by Nic-Ma** [comment]
> > Hi @wyli ,
> > Thanks for your review and feedback. Sorry maybe I didn't describe the PR clearly enough, it doesn't conflict with the `lazy instantiation` idea, it's only the utility to build one component from the dictionary. For the `lazy instantiation` idea, I think it's the logic of handling the **whole config content**, when it's really necessary to create an instance, it will call this utility to create it. I will implement something like `ConfigResolver` to handle the whole config for the step 3 of #3482 later when we discussed with more team guys.
> > Thanks.
> 
> Maybe there's a misunderstanding here, the lazy instantiation will work at the component level as well. Also, it'll be useful to have a complete prototype and then we divide it into smaller pieces and merge. That'll be an efficient way to roll out a major feature.

OK, sounds good to me.
Let's try to discuss and align with team members on the first step of #3482 (the MVP model package draft structure: https://github.com/Project-MONAI/tutorials/pull/487), then I will develop a complete prototype in a separate branch.

Thanks.

File: tests/test_config_parser.py (hunk 1 above)
Comment (rijobro) [reply tests/test_config_parser.py]: Great, thanks!

**Pull request merged**